pkg/ccl/streamingccl/logical/lww_row_processor_test.go

```diff
@@ -0,0 +1,118 @@
+	var key roachpb.Key
+	var numInjectedErrors atomic.Int64
+	injectedErr := kvpb.NewError(errors.New("injected error"))
+
+	ctx := context.Background()
+	srv, sqlDB, kvDB := serverutils.StartServer(b, base.TestServerArgs{
+		DefaultTestTenant: base.TestControlsTenantsExplicitly,
+		Knobs: base.TestingKnobs{
+			Store: &kvserver.StoreTestingKnobs{
+				TestingRequestFilter: func(_ context.Context, ba *kvpb.BatchRequest) *kvpb.Error {
+					// In order to focus the benchmark on the SQL overhead of
+					// evaluating INSERT stmts via the internal executor, we'll
+					// inject errors for them so that the processing of the KV
+					// request stops once it reaches the KV client.
+					if !discardWrite.Load() || !ba.IsWrite() || len(ba.Requests) != 1 {
+						return nil
+					}
+					req, ok := ba.GetArg(kvpb.Put)
+					if !ok {
+						return nil
+					}
+					if !req.(*kvpb.PutRequest).Key.Equal(key) {
+						return nil
+					}
+					numInjectedErrors.Add(1)
+					return injectedErr
+				},
+			},
+		},
+	})
+	defer srv.Stopper().Stop(ctx)
+	s := srv.ApplicationLayer()
+
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+	tableName := "tab"
+	runner.Exec(b, "CREATE TABLE tab (pk int primary key, payload string)")
+	runner.Exec(b, "ALTER TABLE tab ADD COLUMN crdb_internal_origin_timestamp DECIMAL NOT VISIBLE DEFAULT NULL ON UPDATE NULL")
+
+	desc := desctestutils.TestingGetPublicTableDescriptor(kvDB, s.Codec(), "defaultdb", tableName)
+	tableDescs := make(map[string]descpb.TableDescriptor, 1)
+	tableDescs["defaultdb.public."+tableName] = *desc.TableDesc()
+	rp, err := makeSQLLastWriteWinsHandler(ctx, s.Codec(), s.ClusterSettings(), tableDescs)
+	require.NoError(b, err)
+
+	// Insert a single row into the table which we'll attempt to overwrite via
+	// the row processor.
+	// TODO(yuzefovich): I don't quite understand why this is needed - it
+	// appears as if the BatchRequest produced for the rp's INSERT query changes
+	// without it.
+	runner.Exec(b, "INSERT INTO tab VALUES (1, 'hello')")
+
+	// We'll be simulating processing the same INSERT over and over in the loop.
+	keyValue := replicationtestutils.EncodeKV(b, s.Codec(), desc, 1, "hello")
+	keyValue.Value.Timestamp = hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+	key = keyValue.Key
+
+	defer discardWrite.Store(false)
+	discardWrite.Store(true)
+	b.ResetTimer()
+	b.ReportAllocs()
+	var lastTxnErr, lastRowErr error
+	var i int
+	for i = 0; i < b.N; i++ {
+		// We need to start a fresh Txn since we're injecting an error.
+		lastTxnErr = s.InternalDB().(isql.DB).Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+			// The error is expected here due to injected error.
+			lastRowErr = rp.ProcessRow(ctx, txn, keyValue)
+			keyValue.Value.Timestamp.WallTime += 1
+			return nil
+		})
+	}
+	require.Error(b, lastTxnErr)
+	require.Error(b, lastRowErr)
+	// Sanity check that an error was injected on each iteration.
+	require.Equal(b, numInjectedErrors.Load(), int64(i))
```

Comment:
Out of curiosity, I printed lastTxnErr and lastRowErr and got something a bit unexpected:

```
        last txn err: txn already encountered an error; cannot be used anymore (previous err: received transaction in response with earlier read timestamp than in the request. ba.Txn: "unnamed" meta={id=5e85f402 key=/Table/104/1/1/0 iso=Serializable pri=0.02142251 epo=0 ts=1718318491.434912000,0 min=1718318491.434912000,0 seq=0} lock=true stat=PENDING rts=1718318491.434912000,0 wto=false gul=1718318491.934912000,0, br.Txn: "unnamed" meta={id=df05cbce key=/Table/104/1/1/0 iso=Serializable pri=0.04413147 epo=0 ts=1718318490.486129000,0 min=1718318490.486129000,0 seq=0} lock=true stat=PENDING rts=1718318490.486129000,0 wto=false gul=1718318490.986129000,0)
        last row err: replicated-insert: received transaction in response with earlier read timestamp than in the request. ba.Txn: "unnamed" meta={id=5e85f402 key=/Table/104/1/1/0 iso=Serializable pri=0.02142251 epo=0 ts=1718318491.434912000,0 min=1718318491.434912000,0 seq=0} lock=true stat=PENDING rts=1718318491.434912000,0 wto=false gul=1718318491.934912000,0, br.Txn: "unnamed" meta={id=df05cbce key=/Table/104/1/1/0 iso=Serializable pri=0.04413147 epo=0 ts=1718318490.486129000,0 min=1718318490.486129000,0 seq=0} lock=true stat=PENDING rts=1718318490.486129000,0 wto=false gul=1718318490.986129000,0
```

I'm not really sure what is up there. Might not be worth digging into in this PR I suppose.